### PR TITLE
Support Hawei Cloud (PFS && OBS)

### PR DESCRIPTION
**Pull request by yabola**
1. Support Huawei Cloud PFS( parallel file system )
2. Move Huawei Cloud OBS from Alluxio-extension. https://github.com/Alluxio/alluxio-extensions/pull/32
3. fix some UT && add doc for OBS/PFS

PFS is an extension of OBS. 

1. The main difference is  dir in PFS will be explicitly created . But other storage (like s3,obs....) won't create a real dir.
(when PFS,  `OBSUnderFileSystem#getObjectStatus`will return meta in dir , but we want null.).
This will cause dir be considered as a file in some conditions and files in this dir won't be listed.

2.  Assuming this scenario: 
    `/xx/file1/` ( File1 actually exists, which is a file) , but we should pay attention, there is `/` after file1 name.
     When OBS , the path object meta is null.
     When PFS, the path object meta is not null. The object meta is same as `/xx/file1`
     
3.  PFS supports `Rename` directory , OBS doesn't support.

4. Assuming this scenario: 
   There are several files as follows: /dir1, /dir1/file1
   When OBS (also s3...) ,   delete `/dir1` ---> dir1 & file1 all exists.   Then delete `/dir1/file1`, dir1 & file1 all are deleted. 
   When PFS ,  delete `/dir1` ---> dir1 & file1 all exists.   Then delete `/dir1/file1`, file1 is deleted, dir1 is not deleted.
   This means that when PFS, We should first  delete `/dir1/file1`, then delete `/dir1`.

### CODE REVIEW

File: underfs/obs/src/main/resources/META-INF/services/alluxio.underfs.UnderFileSystemFactory [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+# (the “License”). You may not use this work except in compliance with the License, which is
+# available at www.apache.org/licenses/LICENSE-2.0
+#
+# This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+# either express or implied, as more fully set forth in the License.
+#
+# See the NOTICE file distributed with this work for information regarding copyright ownership.
+#
+
+alluxio.underfs.obs.OBSUnderFileSystemFactory
```
Comment (LuQQiu) [inline underfs/obs/src/main/resources/META-INF/services/alluxio.underfs.UnderFileSystemFactory:12]: add an extra line at the end of file

**Review by LuQQiu — COMMENTED** [review]
@yabola can you help modify the OBS doc and possible add PFS doc?
Can you help fix the test failures?

Appreciate your contribution to PFS/OBS!

**Comment by yabola** [comment]
retest this please

**Comment by yabola** [comment]
> @yabola can you help modify the OBS doc and possible add PFS doc?
> Can you help fix the test failures?
> 
> Appreciate your contribution to PFS/OBS!

It seems Jenkins can’t pull https://repo.huaweicloud.com/repository/maven/huaweicloudsdk . How can I retest this in Jenkins?

**Comment by yabola** [comment]
alluxio-bot, test this please

**Comment by yabola** [comment]
@LuQQiu 
I have added the UT and modified related documents.
But It seems Jenkins can’t pull https://repo.huaweicloud.com/repository/maven/huaweicloudsdk

**Comment by LuQQiu** [comment]
@yabola Yeah it cannot pull Huawei dependencies. if it cannot pull, we cannot put it in the main repo since it will prevent us from building the release tarballs. If there are no other ways of getting the dependencies, we can only leave it in the alluxio-extension github repo and hope that users are able to get the dependency when they trying to build the jar

**Comment by yabola** [comment]
> @LuQQiu I found another maven dependency SDK for OBS client (provided by huawei ). It had passed in Jenkins.
> 
> `Failed to execute goal com.mycila:license-maven-plugin:4.0.rc1:check (default) on project alluxio-underfs-obs: Some files do not have the expected license header -> [Help 1]`
> I don't know the reason for this.

Warning: 472 [WARNING] Missing header in: /usr/src/alluxio/underfs/obs/pom.xml
@LuQQiu  I had fixed it.

**Comment by LuQQiu** [comment]
@yabola That's amazing! Thanks for fixing it!

File: docs/en/ufs/OBS.md [hunk 2]
```diff
@@ -76,6 +55,13 @@ Valid values are similar to `obs.cn-east-2.myhuaweicloud.com` and `obs.cn-south-
 Available endpoints are listed in the [Region and Endpoint list](https://developer.huaweicloud.com/en-us/endpoint)
 under the Object Storage Service category.
 
+If you want to use the `PFS` parallel file system, you need to configure in `conf/alluxio-site.properties` as below. The rest of the operations are the same as the `OBS` file system. Add:
```
Comment (LuQQiu) [inline docs/en/ufs/OBS.md:58]: you need to configure in `conf/alluxio-site.properties` as below.
-> please add the following configuration in `conf/alluxio-site.properties`.

File: underfs/obs/src/test/java/alluxio/underfs/obs/OBSOutputStreamTest.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+  /**
+   * Tests to ensure {@link OBSOutputStream#write(byte[], int, int)} calls
+   * {@link OutputStream#write(byte[], int, int)} .
+   */
+  @Test
+  @PrepareForTest(OBSOutputStream.class)
+  public void testWrite2() throws Exception {
+    PowerMockito.whenNew(BufferedOutputStream.class)
+            .withArguments(Mockito.any(DigestOutputStream.class)).thenReturn(mLocalOutputStream);
```
Comment (LuQQiu) [inline underfs/obs/src/test/java/alluxio/underfs/obs/OBSOutputStreamTest.java:111]: please fix the indentations, thanks

File: underfs/obs/pom.xml [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-shade-plugin</artifactId>
+            </plugin>
+            <plugin>
+                <groupId>com.coderplus.maven.plugins</groupId>
+                <artifactId>copy-rename-maven-plugin</artifactId>
+            </plugin>
+        </plugins>
+    </build>
+</project>
```
Comment (LuQQiu) [inline underfs/obs/pom.xml:75]: add a new line at the end of file

File: underfs/obs/src/main/java/alluxio/underfs/obs/OBSPropertyKey.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+import alluxio.conf.PropertyKey;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * OBS configuration property keys.
+ */
+@ThreadSafe
+public final class OBSPropertyKey {
```
Comment (LuQQiu) [inline underfs/obs/src/main/java/alluxio/underfs/obs/OBSPropertyKey.java:22]: Please move all the property key into PropertyKey file and delete this file.
This file is used in ufs-extensions but not needed when move back to Alluxio

**Review by LuQQiu — COMMENTED** [review]
Thanks for moving it back to main repo! left some comments
Please redo the runUfsTests for both OBS and PFS, thanks!

**Comment by LuQQiu** [comment]
@apc999 FYI

**Comment by yabola** [comment]
@LuQQiu  I  had retested `./bin/alluxio runTests` and `runUfsTests` script in both OBS and PFS. Only failed at `getModTimeTest` ( I am using Huawei Cloud in Thailand).

File: underfs/obs/src/main/java/alluxio/underfs/obs/OBSPropertyKey.java (hunk 5 above)
Comment (yabola) [reply underfs/obs/src/main/java/alluxio/underfs/obs/OBSPropertyKey.java]: done

File: underfs/obs/src/main/java/alluxio/underfs/obs/OBSInputStream.java [hunk 6]
```diff
@@ -58,29 +66,29 @@
   /**
    * Creates a new instance of {@link OBSInputStream}.
    *
-   * @param bucketName the name of the bucket
-   * @param key the key of the file
-   * @param client the client for OSS
-   * @param retryPolicy retry policy in case the key does not exist
+   * @param bucketName          the name of the bucket
+   * @param key                 the key of the file
```
Comment (LuQQiu) [inline underfs/obs/src/main/java/alluxio/underfs/obs/OBSInputStream.java:70]: revert the changes?

File: underfs/obs/src/main/java/alluxio/underfs/obs/OBSInputStream.java [hunk 7]
```diff
@@ -58,29 +66,29 @@
...
+   * @param client              the client for OSS
+   * @param retryPolicy         retry policy in case the key does not exist
    * @param multiRangeChunkSize the chunk size to use on this stream
    */
   OBSInputStream(String bucketName, String key, ObsClient client, RetryPolicy retryPolicy,
-      long multiRangeChunkSize) throws IOException {
+                 long multiRangeChunkSize) throws IOException {
```
Comment (LuQQiu) [inline underfs/obs/src/main/java/alluxio/underfs/obs/OBSInputStream.java:76]: revert the changes?

File: underfs/obs/src/main/java/alluxio/underfs/obs/OBSInputStream.java [hunk 8]
```diff
@@ -58,29 +66,29 @@
...
    * Creates a new instance of {@link OBSInputStream}.
    *
-   * @param bucketName the name of the bucket
-   * @param key the key of the file
-   * @param client the OBS client
-   * @param position the position to begin reading from
-   * @param retryPolicy retry policy in case the key does not exist
+   * @param bucketName          the name of the bucket
+   * @param key                 the key of the file
+   * @param client              the OBS client
```
Comment (LuQQiu) [inline underfs/obs/src/main/java/alluxio/underfs/obs/OBSInputStream.java:85]: revert?

File: underfs/obs/src/main/java/alluxio/underfs/obs/OBSInputStream.java [hunk 9]
```diff
@@ -58,29 +66,29 @@
...
+   * @param position            the position to begin reading from
+   * @param retryPolicy         retry policy in case the key does not exist
    * @param multiRangeChunkSize the chunk size to use on this stream
    */
   OBSInputStream(String bucketName, String key, ObsClient client, long position,
-      RetryPolicy retryPolicy, long multiRangeChunkSize) throws IOException {
+                 RetryPolicy retryPolicy, long multiRangeChunkSize) throws IOException {
```
Comment (LuQQiu) [inline underfs/obs/src/main/java/alluxio/underfs/obs/OBSInputStream.java:91]: revert?

File: underfs/obs/src/main/java/alluxio/underfs/obs/OBSOutputStream.java [hunk 10]
```diff
@@ -121,7 +135,7 @@ public void write(byte[] b) throws IOException {
    * Writes the given number of bytes from the given byte array starting at the given offset to this
    * output stream. Before close, the bytes are all written to local file.
    *
-   * @param b the byte array
+   * @param b   the byte array
```
Comment (LuQQiu) [inline underfs/obs/src/main/java/alluxio/underfs/obs/OBSOutputStream.java:138]: revert?

File: underfs/obs/src/main/java/alluxio/underfs/obs/OBSUnderFileSystem.java [hunk 11]
```diff
@@ -49,36 +49,44 @@
...
-        "Property %s is required to connect to OBS", OBSPropertyKey.OBS_ACCESS_KEY);
-    Preconditions.checkArgument(conf.isSet(OBSPropertyKey.OBS_SECRET_KEY),
-        "Property %s is required to connect to OBS", OBSPropertyKey.OBS_SECRET_KEY);
-    Preconditions.checkArgument(conf.isSet(OBSPropertyKey.OBS_ENDPOINT),
-        "Property %s is required to connect to OBS", OBSPropertyKey.OBS_ENDPOINT);
-    String accessKey = conf.get(OBSPropertyKey.OBS_ACCESS_KEY);
-    String secretKey = conf.get(OBSPropertyKey.OBS_SECRET_KEY);
-    String endPoint = conf.get(OBSPropertyKey.OBS_ENDPOINT);
-    String bucketType = conf.get(OBSPropertyKey.OBS_BUCKET_TYPE);
+                                                  UnderFileSystemConfiguration conf) throws Exception {
```
Comment (LuQQiu) [inline underfs/obs/src/main/java/alluxio/underfs/obs/OBSUnderFileSystem.java]: Indentation?

File: underfs/obs/src/main/java/alluxio/underfs/obs/OBSOutputStream.java [hunk 12]
```diff
@@ -46,33 +46,47 @@
...
-   * @param key the key of the file
-   * @param client the OBS client
-   * @param tmpDirs a list of temporary directories
+   * @param key        the key of the file
+   * @param client     the OBS client
+   * @param tmpDirs    a list of temporary directories
    */
   public OBSOutputStream(String bucketName, String key, ObsClient client,
-      List<String> tmpDirs) throws IOException {
+                         List<String> tmpDirs) throws IOException {
```
Comment (LuQQiu) [inline underfs/obs/src/main/java/alluxio/underfs/obs/OBSOutputStream.java:89]: Revert?

**Review by LuQQiu — APPROVED** [review]
Left some comments about code style
our indentation is 2 spaces, for separating line it's 4 spaces
Please fix the codestyle issue, otherwise LGTM
thanks for testing out

**Comment by yabola** [comment]
@LuQQiu Thanks for the detailed review. I change my IDE configuration ( 
`Method declaration parameters` ->  `align when multiline`) and reformat my code.

File: underfs/obs/src/main/java/alluxio/underfs/obs/OBSConstants.java [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+package alluxio.underfs.obs;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * OBS constants.
+ */
+@ThreadSafe
+public final class OBSConstants {
+  public static final String HEADER_OBS = "obs://";
```
Comment (LuQQiu) [inline underfs/obs/src/main/java/alluxio/underfs/obs/OBSConstants.java:21]: Sorry one more thing!
Can you remove this file and put this header into `alluxio.Constants` together with other ufs like s3://

**Review by LuQQiu — COMMENTED** [review]
Left one minor comments, otherwise this PR is ready to be merged

File: underfs/obs/src/main/java/alluxio/underfs/obs/OBSConstants.java (hunk 13 above)
Comment (yabola) [reply underfs/obs/src/main/java/alluxio/underfs/obs/OBSConstants.java]: done~

**Comment by LuQQiu** [comment]
alluxio-bot, merge this please

**Comment by LuQQiu** [comment]
@yabola can you open a new PR to give more information about what is PFS VS OBS in alluxio documentation?

**Comment by yabola** [comment]
> @yabola can you open a new PR to give more information about what is PFS VS OBS in alluxio documentation?

@LuQQiu Ok~ What points need to be said? like PR description?
